llvm/test/CodeGen/X86/avgfloors-scalar.ll

```diff
@@ -167,11 +151,11 @@ define i32 @test_ext_i32(i32 %a0, i32 %a1) nounwind {
 ;
 ; X64-LABEL: test_ext_i32:
 ; X64:       # %bb.0:
-; X64-NEXT:    movl %edi, %eax
-; X64-NEXT:    andl %esi, %eax
-; X64-NEXT:    xorl %esi, %edi
-; X64-NEXT:    sarl %edi
-; X64-NEXT:    addl %edi, %eax
+; X64-NEXT:    movslq %esi, %rcx
+; X64-NEXT:    movslq %edi, %rax
+; X64-NEXT:    addq %rcx, %rax
+; X64-NEXT:    shrq %rax
+; X64-NEXT:    # kill: def $eax killed $eax killed $rax
```

Comment:
Maybe I'm missing something, but shouldn't we have `sarq` here for the signed version?

Reply:
Yes that scared me as well at first :) We have excess sign bits (32 from i32->i64 sext then add) , and we're truncating the result back to i32, so SimplifyDemandedBits replaces the i64 sra with i64 srl.

Reply:
So we can just replace `ShiftOpc` with `ISD::SRL` in the code?